### PR TITLE
Migrate to Chainer v2 

### PR DESCRIPTION
**Pull request by Hakuyume**
### upgrades
https://docs.chainer.org/en/v2.0.0/upgrade.html
- [x] update config files `1.24` -> `2.0`
- [x] `volatile` -> `with chainer.function.no_backprop_mode()`
    https://docs.chainer.org/en/v2.0.0/upgrade.html#volatile-flag-is-removed
- [x] `use_cudnn` -> (remove)
    https://docs.chainer.org/en/v2.0.0/upgrade.html#configurations-are-added-and-replace-some-of-existing-global-flags
- [x] `train`/`test` -> `chainer.config.train`
    https://docs.chainer.org/en/v2.0.0/upgrade.html#training-mode-is-configured-by-a-thread-local-flag
- [x] new-style parameter/child link registration
    https://docs.chainer.org/en/v2.0.0/upgrade.html#new-style-parameter-registration-apis-are-added-to-link
    https://docs.chainer.org/en/v2.0.0/upgrade.html#new-style-child-link-registration-apis-are-added-to-chain
- [x] `invoke_before_training` -> (remove)
    https://docs.chainer.org/en/v2.0.0/upgrade.html#extension-invoke-before-training-is-removed
- [x] `(None, out_size)` -> `out_size`
    https://docs.chainer.org/en/v2.0.0/upgrade.html#update-omit-input-size

### tests
- [x] nosetests with CPU/GPU
- [x] `examples/detection/visualize_models.py` with CPU
- [x] `examples/detection/eval_voc07.py` with GPU
    - [x] Faster R-CNN
    - [x] SSD300
    - [x] SSD512
- [x] `examples/segnet/evaluate.py` with GPU

### CODE REVIEW

**Comment by yuyu2172** [comment]
How about using this.
https://docs.chainer.org/en/v2.0.0/upgrade.html#new-style-parameter-registration-apis-are-added-to-link

**Comment by yuyu2172** [comment]
Also, `invoke_before_training` should be deleted.
https://github.com/chainer/chainercv/blob/master/chainercv/extensions/detection/detection_vis_report.py#L76

https://docs.chainer.org/en/v2.0.0/upgrade.html#extension-invoke-before-training-is-removed

**Comment by Hakuyume** [comment]
@yuyu2172 Thank you for pointing out.

File: chainercv/links/model/segnet/segnet_basic.py [hunk 1]
```diff
@@ -133,23 +138,23 @@ def __call__(self, x):
...
+        p1 = F.MaxPooling2D(2, 2)
+        p2 = F.MaxPooling2D(2, 2)
+        p3 = F.MaxPooling2D(2, 2)
+        p4 = F.MaxPooling2D(2, 2)
         h = F.local_response_normalization(x, 5, 1, 1e-4 / 5., 0.75)
-        h = p1(F.relu(self.conv1_bn(self.conv1(h), test=not self.train)))
-        h = p2(F.relu(self.conv2_bn(self.conv2(h), test=not self.train)))
-        h = p3(F.relu(self.conv3_bn(self.conv3(h), test=not self.train)))
-        h = p4(F.relu(self.conv4_bn(self.conv4(h), test=not self.train)))
+        h = _without_cudnn(p1, F.relu(self.conv1_bn(self.conv1(h))))
```
Comment (yuyu2172) [inline chainercv/links/model/segnet/segnet_basic.py]: How about.
I am not sure what is the best practice to do this, but I think the code below looks easier to read.

```python
with chainer.using_config('use_cudnn', 'never'):
    h = p1(F.relu(....))
    h = p2(F.relu(....))
    h = p3(F.relu(....))
    h = p4(F.relu(....))

```

File: tests/links_tests/model_tests/faster_rcnn_tests/test_region_proposal_network.py [hunk 2]
```diff
@@ -35,10 +35,12 @@ def setUp(self):
         self.x = np.random.uniform(size=(self.B, C, H, W)).astype(np.float32)
         self.img_size = (W * feat_stride, H * feat_stride)
 
-    def _check_call(self, x, img_size, test):
+        chainer.config.train = not self.test
```
Comment (yuyu2172) [inline tests/links_tests/model_tests/faster_rcnn_tests/test_region_proposal_network.py]: https://github.com/Hakuyume/chainercv/blob/df0fe8a608835a8de2a7d8c731e82bdf96d5cd08/tests/links_tests/model_tests/faster_rcnn_tests/test_region_proposal_network.py#L13

How about `self.test --> self.train`?

Comment (yuyu2172) [inline tests/links_tests/model_tests/faster_rcnn_tests/test_region_proposal_network.py]: Also test_proposal_creator needs to be changed.

File: chainercv/links/model/segnet/segnet_basic.py (hunk 1 above)
Comment (Hakuyume) [inline chainercv/links/model/segnet/segnet_basic.py]: At first, I wrote as you suggested. But I noticed convolution2ds were executed without `cudnn`. I thought this was not good for performance.

Comment (yuyu2172) [inline chainercv/links/model/segnet/segnet_basic.py]: You are right.
I missed that.
lgtm

**Comment by yuyu2172** [comment]
LGTM

**Pull request merged**